wrapper.py

```diff
@@ -0,0 +1,148 @@
+
+    os.makedirs(logdir, exist_ok=True)
+
+    logfile = os.path.join(
+        logdir,
+        "QD-rnaseq-wrapper_" + now.strftime("%y%m%d_%H%M%S") + ".log",
+        )
+    logger = setup_logger("qd-rnaseq", logfile)
+
+    #### --- Read in the file with previously analysed samples --- ###
+    previous_samples = read_previous_samples_file(config)
+
+    ### --- Find all slims records marked for QD-RNAseq pipeline as secondary analysis --- ###
+    rnaseq_samples = samples_from_sec_analysis(186)
+    # 29 = WOPR
+    # 186 = QD-RNA
+
+    # Skip the sample if it has previously been analysed
+    # This also gathers info from SLIMS about the sample
+    # This is a bit cumbersome but done as SLIMS doesn't have the run tag easily queryable right now
+    for sample in list(rnaseq_samples.keys()):
+        # Get the runtag for the sample, combine with sample name
+        try:
+            slimsinfo = SlimsSample(sample)
+            runtag = find_runtag_from_fastqs(slimsinfo.fastqs)
+            sample_id = f"{sample}_{runtag}"
+        except Exception as e:
+            logger.error(e)
+            sys.exit(1)
+            # TODO, this should mark this sample as failed, and send an email. Not break the whole thing
+
+        if sample_id in previous_samples:
+            del rnaseq_samples[sample]
+
+    # Check if any samples remains, if so start the pipeline
+    if len(rnaseq_samples) > 0:
+        logger.info(f"Found {len(rnaseq_samples)} sample(s) marked for QD-RNAseq pipeline. Starting the wrapper.")
+    else:
+        sys.exit(0)
+
+    ### --- Loop over each record --- ###
+    runner_samples = {} # Store all samplesheet paths per sample in a dict for later use
+    for sample, record in rnaseq_samples.items():
+        ### --- Collect information about the sample --- ###
+        # Create a new SlimsSample object
+        logger.info(f"{sample} - Extracting SLIMS information.")
+        slimsinfo = SlimsSample(sample)
+
+        # Translate the information from the SLIMS database into a dictionary
+        info = translate_slims_info(record)
+
+        # Get the run tag and combine with name
+        runtag = find_runtag_from_fastqs(slimsinfo.fastqs)
+        sample_id = f"{sample}_{runtag}"
+
+        ### --- Generate a samplesheet from the information gathered --- ###
+        outdir = os.path.join(config.get("general", "output_dir"), sample_id)
+        logger.info(f"{sample} - Generating samplesheet.csv in {outdir}.")
+        os.makedirs(outdir, exist_ok=True)
+
+        strandedness = config.get("general", "strandedness")
+
+        sample_ss_path = make_samplesheet(sample_id, slimsinfo.fastqs, strandedness, outdir)
+        runner_samples[sample_id] = sample_ss_path
+
+    ### --- Start a runner for each sample --- ###
+    completed_samples = start_runner_threads(runner_samples, logger)
+
+    ### --- Add analysed samples to previously analysed samples file --- ###
+    for sample in completed_samples:
```

Comment:
Will this suffer from potential race conditions where a secondary run with the wrapper may pick up an already running sample? Say a process is somehow stuck in the grid engine queue when the wrapper starts going a second time?

Reply:
Yes, totally agree.

But instead of fixing this here, I'd say that this is better fixed by using secondary analysis run states in SLIMS in a coming PR